### ISSUE TITLE
Support T:?Sized as type parameter for Mutex and other types

### ISSUE THREAD

**Issue by StevenJiang1110** [asterinas/asterinas#738]
# Background

Currently, The `Mutex` and `SpinLock` in Asterinas can only accept sized type.

```rust
pub struct Mutex<T> {
    // ....
}
```

While for these types in [spin crate](https://docs.rs/spin/0.9.8/spin/mutex/struct.MutexGuard.html) and [std](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html), they can both support dynamic-sized type.

```rust
pub struct Mutex<T: ?Sized> {
    // ....
}
```
We hope to implement the similar behavior for `Mutex` and `SpinLock` in Asterinas.

# Benefits

If these lock types can accept `T:?Sized`(such as `&dyn Trait` and `[T]`) as the type parameter, we can simplify multiple usages in current codebase. Below are several examples.

```rust
// original: Arc<SpinLock<Box<dyn AnyNetworkDevice>>>
// new: Arc<SpinLock<dyn AnyNetworkDevice>>

// original: Once<Arc<Mutex<Box<dyn Apic + 'static>>>>
// new: Once<Arc<Mutex<dyn Apic + 'static>>>
```

# Design

TODO

# Drawbacks

To support unsized type, `Mutex` should store the size of type. It may occupy a little bit more memory (8 bytes for each lock)

# Future possibilities

Not only lock types, but other types which can be viewed as fat pointers should provide similar behavior as existing fat pointers.

**Comment by StevenJiang1110** [comment]
This change may not be at a high priority and requires a review process.

**Comment by tatetian** [comment]
This feature request makes sense to me and the implementation seems to be straightforward. @grief8 Could you take this task?

**Comment by grief8** [comment]
> This feature request makes sense to me and the implementation seems to be straightforward. @grief8 Could you take this task?

Of course! I am willing to do it.

**Comment by grief8** [comment]
I have added `T: ?Sized` as type parameter in Mutex, SpinLock, RwLock, and RwMutex and replaced old expressions with simplified usages to prove the correctness of changes in #753 . 

For example,

```
- pub static APIC_INSTANCE: Once<Arc<Mutex<Box<dyn Apic + 'static>>>> = Once::new();
+ pub static APIC_INSTANCE: Once<Arc<Mutex<dyn Apic + 'static>>> = Once::new();

- driver: Arc<SpinLock<Box<dyn AnyNetworkDevice>>>,
+ driver: Arc<SpinLock<dyn AnyNetworkDevice>>,
```

**Comment by StevenJiang1110** [comment]
Fixed by #753.
